core/src/main/scala/cats/data/Chain.scala

```diff
@@ -574,6 +574,62 @@ sealed abstract class Chain[+A] {
    */
   final def size: Long = length
 
+  /**
+   * The number of elements in this chain, if it can be cheaply computed, -1 otherwise.
+   * Cheaply usually means: Not requiring a collection traversal.
+   */
+  final def knownSize: Long =
+    this match {
+      case _ if isEmpty       => 0
+      case Chain.Singleton(_) => 1
+      case _                  => -1
+    }
+
+  /**
+   * Compares the length of this chain to a test value.
+   * 
+   * The method does not call `length` directly; its running time
+   * is `O(length min len)` instead of `O(length)`.
+   * 
+   * @param  len the test value that gets compared with the length.
+   * @return a negative value if `this.length < len`,
+   *         zero if `this.length == len` or
+   *         a positive value if `this.length > len`.
+   * @note   This is an adapted version of
+             [[https://github.com/scala/scala/blob/v2.13.8/src/library/scala/collection/Iterable.scala#L272-L288 Iterable#sizeCompare]]
+             from Scala Library v2.13.8
+   * 
+   * {{{
+   * scala> import cats.data.Chain
+   * scala> val chain = Chain(1, 2, 3)
+   * scala> val isLessThan4 = chain.lengthCompare(4) < 0
+   * scala> val isEqualTo3 = chain.lengthCompare(3) == 0
+   * scala> val isGreaterThan2 = chain.lengthCompare(2) > 0
+   * scala> isLessThan4 && isEqualTo3 && isGreaterThan2
+   * res0: Boolean = true
+   * }}}
+   */
+  final def lengthCompare(len: Long): Int =
```

Comment:
I wonder about an alternate approach:
```scala
// invariant, sz >= 0
final def lengthCompare(len: Int): Int =
  this match {
    case Empty => Integer.compare(0, len)
    case Singleton(_) => Integer.compare(1, len)
    case Wrap(seq) => seq.lengthCompare(len)
    case _ =>
      // length 2 or more... iterate through
}
```

Since this would handle a some common cases without allocation: empty, singletons and chains that are wrapping seqs (which are returned by several operations because they are more efficient).

Reply:
Yes sure – we can add more optimizations here. The only note: `Chain` uses `Long` for length/size measurements. Do you know why, by chance?

Reply:
Added some optimized special cases. I think I may need to add more stress on those cases in tests.